### PR TITLE
#2469 Adding getTimeUntilDisconnectNotification() to DownstreamMessage

### PR DESCRIPTION
**Pull request by fkaltner**
This is done as preparation for adapting ITs to Kafka. When ITs are based upon DownstreamMessage, `DownstreamMessage#getTimeUntilDisconnectNotification ` can be used to obtain the time until disconnect notification instead of `TimeUntilDisconnectNotification.fromMessage(msg)`.

### CODE REVIEW

File: clients/application/src/main/java/org/eclipse/hono/application/client/DownstreamMessage.java [hunk 1]
```diff
@@ -123,4 +125,31 @@ default boolean isSenderConnected(final Instant now) {
...
+            return Optional.empty();
+        } else if (ttd == 0 || MessageHelper.isDeviceCurrentlyConnected(ttd,
+                creationTime != null ? creationTime.toEpochMilli() : null)) {
+            final String tenantId = getTenantId();
+            final String deviceId = getDeviceId();
+
+            if (tenantId != null && deviceId != null) {
+                final TimeUntilDisconnectNotification notification =
+                        new TimeUntilDisconnectNotification(tenantId, deviceId, ttd,
+                                TimeUntilDisconnectNotification.getReadyUntilInstantFromTtd(ttd, creationTime), creationTime);
```
Comment (sophokles73) [inline clients/application/src/main/java/org/eclipse/hono/application/client/DownstreamMessage.java:148]: it looks very starnge to pass in ttd and creation time as parameters and then also being required to use the same parameters to compute another one. Shouldn't the constructor take care of that?

Comment (fkaltner) [reply clients/application/src/main/java/org/eclipse/hono/application/client/DownstreamMessage.java:148]: Absolutely! It's implemented in `TimeUntilDisconnectNotification#fromMessage` which I used as a template, I'll fix it there as well.

File: clients/application/src/main/java/org/eclipse/hono/application/client/DownstreamMessage.java [hunk 2]
```diff
@@ -123,4 +125,30 @@ default boolean isSenderConnected(final Instant now) {
...
+     * @return A notification if the message contains a TTD value {@link Optional#empty()} otherwise.
+     */
+    default Optional<TimeUntilDisconnectNotification> getTimeUntilDisconnectNotification() {
+
+        final Integer ttd = getTimeTillDisconnect();
+        final Instant creationTime = getCreationTime();
+
+        if (ttd == null) {
+            return Optional.empty();
+        } else if (ttd == 0 || MessageHelper.isDeviceCurrentlyConnected(ttd,
```
Comment (sophokles73) [inline clients/application/src/main/java/org/eclipse/hono/application/client/DownstreamMessage.java:140]: I'd rather not introduce this dependency on *MessgeHelper* here as it is basically a helper class for dealing with AMQP messages and pulls in a lot of transitive dependencies. Maybe we can encapsulate the required logic in *TimeUntilDisconnectNotification* as well?

Comment (fkaltner) [reply clients/application/src/main/java/org/eclipse/hono/application/client/DownstreamMessage.java]: Done.

File: core/src/main/java/org/eclipse/hono/util/TimeUntilDisconnectNotification.java [hunk 3]
```diff
@@ -187,4 +185,54 @@ public long getMillisecondsUntilExpiry() {
...
+     *
+     * @param ttd The TTD value.
+     * @param creationTime The creation time of the message. If {@code null} the device is considered as disconnected.
+     * @return {@code true} if the TTD value contained in the message indicates that the device will
+     *         stay connected for some additional time.
+     */
+    public static boolean isDeviceCurrentlyConnected(final Integer ttd, final Long creationTime) {
+
+        return Optional.ofNullable(ttd).map(ttdValue -> {
+            if (ttdValue == MessageHelper.TTD_VALUE_UNLIMITED) {
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/TimeUntilDisconnectNotification.java:205]: Hmmm, same problem, I guess. If we really want to get rid of the AMQP dependencies we will probably need to move the field constants to some more technology-agnostic class, e.g. *Constants*. But I guess we should do that in another PR ...

**Review by sophokles73 — APPROVED** [review]

File: core/src/main/java/org/eclipse/hono/util/TimeUntilDisconnectNotification.java (hunk 3 above)
Comment (fkaltner) [reply core/src/main/java/org/eclipse/hono/util/TimeUntilDisconnectNotification.java:205]: I think, if we really want to get rid of unnecessary dependencies of the client module we need to cut the Maven dependency to hono-core which is still AMQP-specific.

Comment (sophokles73) [reply core/src/main/java/org/eclipse/hono/util/TimeUntilDisconnectNotification.java:205]: If we only use classes from hono-core which do not depend on e.g. vertx-proton and/or proton-j then we can at least exclude those artifacts from the declared dependency to hono-core. But that is not possible if we need e.g. MessageHelper and MessageHelper uses proton-j ...
